### DISCUSSION TITLE
HowToAdd Sdongle

### DISCUSSION THREAD

**Discussion by Pappnouse** [wlcrs/huawei_solar#466] [General]
What must I do to add a SDONGLE? SDonge has some adresses which I need to add to home assistant. Description of SDonge Modbus can be found here: https://support.huawei.com/enterprise/en/doc/EDOC1100322944?section=k004 (SDonge V200R022CC10) and I tested a connection with Node-RED and can read adress 37498 (Total input power) with device ID 100. 
I also had checked the code sources on this site but can´t really find what is to do to add the sdongle so I need a hint. 
Thx and best regards, Stephan

**Comment by wlcrs** [comment]
You will need to add support in the underlying [`huawei-solar` library](https://gitlab.com/Emilv2/huawei-solar/) first. As I do not have an SDongle myself - and thus cannot test against it - I will not do any work on this. Feel free to create a PR early in the development to get initial feedback!

**Reply by Pappnouse** [reply]
Thx for your fast answer :-) So I will check the underlying project and will come back to this topic.

**Reply by Roving-Ronin** [reply]
@Pappnouse are you trying to use the WiFi/FE  SDongle and this integration to have your Hauwei inverters visible / controllable in Home Assistant ? A bit confusing with your reference to Node-Red...
With this integration there is no need for calling modbus registers with Node-Red, as wlcrs has taken care of all that and it can be done with the integration. If you just having issues getting it installed and communicating with HA, then suggest reading the documentation:
https://github.com/wlcrs/huawei_solar/wiki/Connecting-to-the-inverter  or for cascaded multiple inverters see: https://github.com/wlcrs/huawei_solar/wiki/Connecting-to-daisy-chained-inverters

Personally I am running with SDongle V200R022C10SPC108 and 2 cascaded inverters running the latest V100R001C00SPC155  (there is a SPC156 but it only adds branding for OEM inverters for a distribution network change happening in Australia / New Zealand about to the announced) and this configuration works fine 'out of the box' with this integration using said sdongle (and no Node-Red needed).

**Reply by Pappnouse** [reply]
Hi @Roving-Ronin, 
I would not confusing you :-) I have an 12 years old inverter and the configuration is exaclty like the example hier: https://support.huawei.com/enterprise/en/doc/EDOC1100167258/85f29830/networking-with-third-party-inverters .
With NodeRed I only test if I can read address 37498 (Total input power) no other plan with NodeRed. In the documentation (SDONGE V200R022CC10 Modbus Interface Definitions) I can find the description of address 37498 and I need this information in Home Assistant to display the correct input power of the new and old inverter. The problem (If I am correct) is that address 37498 is on SDONGLE device and Home Accistant only detects one power meter, the inverter and battery. 
Best regards,
Stephan

**Reply by Roving-Ronin** [reply]
So your trying to get the 'Total Input Power' of your Huawei (Model ??) inverter + the "old" inverter ?

If so, then your not going to be able to, as this is a reading of the PV input panel "internal" to each inverter.

According to the Huawei documentation, the Huawei inverter is connected to the 2nd DDSU / DTSU meter via RA485. It has no visibility of the internal metrics of the 'old' inverter. Given this you will only be able to read (from the 2nd meter):
- Active Power W
- Consumption kWh (lifetime of meter)
- Exported kWh (lifetime of meter)
- Frequency Hz
- Power Factor
and Active Power / Voltage / Current / PF of each phase, if 3 phase.

Given the old unit is 12 years old, and the relatively cheap cost of Huawei inverter, is replacing the old inverter with a new Huawei L1 or M1 inverter not an option? If you can, this should be relatively straight forward moving MC4 connectors from PV strings to new inverter, change AC plugs and connect AC to new inverter, then you'd just need to connect RS485 from inverter 1 out pins to inverter 2 in pins.
With that you'll get full visibility of the consumption meter and  both inverters (& any battery), whilst also being able to control the inverters/battery.

**Reply by Pappnouse** [reply]
To your 1st question: SUN2000-5KTL-M1
You are correct I need only the information from the 2nd meter:
Total Input Power - 37498
Load Power - 37500
Total active power - 37516
The measurements from the 2nd meter can only be read out via sdongle.
The old inverter works fine and no replacement is planned.

**Reply by Roving-Ronin** [reply]
@Pappnouse,
What make / model is the 'old' inverter ?

**Reply by Pappnouse** [reply]
The old inverter is SMA Tripower 10000TL-10

**Reply by Roving-Ronin** [reply]
@Pappnouse,
There is no 'use' for this integration to add support for the additional registers that you want, and doing so would only give you a couple of metrics off the 2nd DDSU-666 meter, non of which is the 'Total Input Power' that you want !

Assuming you have the RS485 connectivity for the SMA Tripower 10000TL-10? IMHO I would recommend buying an Elfin EE-11 (Ethernet) or EW-11 (WiFi) appliance for USD$15 from AliExpress, that will give you a Modbus RTU to Modbus TCP device.  Here's the Elfin EW-11 - https://www.aliexpress.com/item/1005002716265770.html

Connecting the SMA via RS485 serial cable to the Elfin modbus bridge (i.e the RTU side), will then allow you to query the SMA over the network (i.e. the TCP side), as your trying to do with the Huawei dongle. This will then seperate the two different inverters and keep the software integration you need for each in HA seperate also.

As it stands currently with the dongle wired to the SMA, you don't have any control over the SMA from the Huawei... so the Huawei can't tell it to throttle down production if you want to limit export for example. If the Huawei and SMA were not connected, the Huawei should still 'see' the power being producted by the SMA (as the Huawei DDSU-666 should be the primary meter with CT located closest to the grid connection). Given this the Huawei would then still be able to utilise the SMA power being produced and charge the any LUNA battery, and throttle itself down, if you needed to reduce production / export limit.

I would suggest doing the following:

1. Setup the Huawei and its DDSU-666 as the primary inverter & meter (with no communication to the SMA), then install the wlcrs integration into HA and add the Huawei inverter, using the network connection option.   At this point you will have the 5kW M1 + DDSU-666 fully functional in HA.

2. Connect the Elfin modbus bridge to the SMA inverter via its RS485 port, and get the Elfin confirmed on your network.  Then in HA try add the SMA integration ( https://www.home-assistant.io/integrations/sma/ )  and see if you can connect to the SMA inverter.  If you can then great, if not (if the SMA integration doesn't support this inverter as its too old) then I would suggest reading:

https://www.home-assistant.io/integrations/modbus/  and https://community.home-assistant.io/t/integration-with-sma-solar-stp-10000tl-20-via-sbfspot/204278

You should be able to get the SMA inverter connected into HA using one of these methods. Using one of these methods, you should also be able to get access to the FULL range of modbus registers ( and the data they provide) for the SMA inverter... including the 'Total Active Power' / 'Load Power' / 'Total Input Power' registers your wanting, and if the SMA integration supports it also able send commands to the SMA inverter (i.e. curtail production / restrict export etc).

3. At this point you should have two seperate integrations providing full access to both inverters, but with their own (PV) Input Power, Active Power etc.  So now to get a combined / total sensor, you'll just need to create a couple of easily made template sensors in HA. So if you want Active Power for 5kW M1 + 10kW SMA you just need to look in HA to find the name/id of these individual sensors, then using the template sensor just update value_template 'formula' to combine the two.

See this for an example, that you could modify for this purpose:  https://github.com/wlcrs/huawei_solar/wiki/Connecting-to-daisy-chained-inverters#combining-multiple-inverter-values

For example your 'Total Input Power' of both inverters would be something like this:

```
template:
  - sensor:
    - name: "Inverters Input Power"
      unique_id: inverters_input_power
      unit_of_measurement: "W"
      device_class: "power"
      state_class: measurement
      state: "{{ (states('sensor.huawei_inverter_input_power') | float) + (states('sensor.sma_inverter_input_power') | float)}}"
```

**Reply by Pappnouse** [reply]
I will do a try with your solution. Thx for this information.
Best regards,
Stephan

**Reply by Pappnouse** [reply]
@Roving-Ronin
Elfin EE11 received but it looks like it doesn´t work with my SMA inverter (Firmware 2.33). I tested as described in this article: https://www.sma-sunny.com/en/how-to-test-the-connection-to-your-sma-inverter/ . I receive a time out via the Elfin bridge :-( 
but... I have a Sunny Webbox and on this Webbox I found an option to enable modbus and voila I receive information from my inverter. 1st in HA was to use SMA integration but as you mentioned my inverter is too old. So next step was to use HA modbus integration, problem here is my HA installation is a installation as a Docker container. After modification of configuration.yaml I receive the error: "Component error: modbus - cannot import name 'Defaults' from 'pymodbus.constants' (/config/deps/lib/python3.11/site-packages/pymodbus/constants.py)" . I found this article: https://github.com/thingsboard/thingsboard-gateway/issues/1158 but the error is still exist :-(  This is the current situation and I am still searching :-)

**Reply by Roving-Ronin** [reply]
You could always pickup a Lenovo Tiny PC and install HA onto that ;-)
Personally I've run HA on a VM and also using Docker (OK for a Dev/Test instance to try updates and config changes upon before installing on the main / production HA instance) and got sick of the slower performance and limitations, such as with Docker you won't get HACS going and connecting things like a USB ZWave dongle is a pain.

M72 Tiny PC with an (iirc) i5-4560 CPU, 16GB RAM cost AUD$120 (approx USD$80) and tossed a spare Samsung 1TB 2.5"SSD into it.. HA blazes along now.

**Reply by Pappnouse** [reply]
@Roving-Ronin  would like to say thanks for your hints :-) The old inverter is now visible in HA. I also find other infomration on this site: https://community.home-assistant.io/t/modbus-data-from-sma-inverter/122425/285 and I have also posted my solution there.
HA found a new home on this hardware: LENOVO ThinkCentre M910 Tiny i5/16GB/512GB NVMe SSD
Best regards,
Stephan

**Reply by Roving-Ronin** [reply]
@Pappnouse I would suggest adding a unique_id to each sensor. Makes it easier longer term if you need to edit in via the GUI, or delete for some reason, without being told it has no unique_id so you can't remove in the GUI. Also means you can use a 'normal' text format for the name, as the sensor will be created with the unique_id.

For example, only doing one, sure you can work out for the others...

```

# SMA Modbus ST10000-10 via Sunny Webbox
modbus:
 - name: SMA
   type: tcp
   host: 192.168.xxx.xx
   port: 502
   sensors:
      - name: "SMA Power AC"
        unique_id: sma_power_ac
        state_class: measurement
        device_class: energy
        unit_of_measurement: W
        slave: 3
        address: 30775
        data_type: uint32
        scan_interval: 60
```

**Reply by Pappnouse** [reply]
@Roving-Ronin you are right. Next step is to make a code review :-)
